DEVINFO.md

```diff
@@ -0,0 +1,174 @@
+
+### Error Handling
+
+* **Each subsystem must have its own `Error` type.**  Error types of aggregate subsystems are encouraged to both wrap their constituent subsystems' `Error` types in their own `Error` types, as well as provide conversions from them via a `From` trait implementation.
+
+* Functions that act on externally-submitted data **must never panic**.  This includes code that acts on incoming network messages, blockchain data, and burnchain (Bitcoin) data.
+
+* **Runtime panics should be used sparingly**.  Generally speaking, a runtime panic is only appropriate if there is no reasonable way to recover from the error condition.  For example, this includes (but is not limited to) disk I/O errors, database corruption, and unreachable code.
+
+* If a runtime panic is desired, it **must have an appropriate error message**.
+
+### Logging
+
+* Log messages should be informative and context-free as possible.  They are used mainly to help us identify and diagnose problems.  They are _not_ used to help you verify that your code works; that's the job of a unit test.
+
+* **DO NOT USE println!() OR eprintln!()**.  Instead, use the logging macros (`test_debug!()`, `trace!()`, `debug!()`, `info!()`, `warn!()`, `error!()`).
+
+* Use **structured logging** whenever you find yourself logging multiple data with a format string.
+
+* Use `trace!()` and `test_debug!()` liberally.  It only runs in tests.
+
+* Use `debug!()` for information that is relevant for diagnosing problems at runtime.  This is off by default, but can be turned on with the `BLOCKSTACK_DEBUG` environment variable.
+
+* Use `info!()` sparingly.
+
+* Use `warn!()` or `error!()` only when there really is a problem.
+
+### Consensus-Critical Code
+
+A **consensus-critical change** is a change that affects how the Stacks blockchain processes blocks, microblocks, or transactions, such that a node with the patch _could_ produce a different state root hash than a node without the patch.  If this is even _possible_, then the PR is automatically treated as a consensus-critical change and must ship as part of a hard fork.  It must also be described in a SIP.
+
+* **All changes to consensus-critical code must be opened against `next`**.  It is _never acceptable_ to open them against `develop` or `master`.
+
+* **All consensus-critical changes must be gated on the Stacks epoch**.  They may only take effect once the system enters a specific epoch (and this must be documented).
+
+A non-exhaustive list of examples of consensus-critical changes include:
+
+* Adding or changing block, microblock, or transaction wire formats
+* Changing the criteria under which a burnchain operation will be accepted by the node
+* Changing the data that gets stored to a MARF key/value pair in the Clarity or Stacks chainstate MARFs
+* Changing the order in which data gets stored in the above
+* Adding, changing, or removing Clarity functions
```

Comment:
The wording is broad on purpose, because this also applies to PRs that attempt to retroactively change the behavior of any deployed smart contract function _or_ built-in function.  The former kind of change is almost certainly going to be rejected, since retroactively altering other peoples' smart contracts is generally frowned upon by the greater ecosystem (why have a blockchain at all if the code isn't immutable?).  The latter is permitted, but it must be part of a consensus-changing SIP and properly feature-gated.